jaeger-core/src/test/java/io/jaegertracing/internal/JaegerSubclassTest.java

```diff
@@ -0,0 +1,231 @@
+                                          ScopeManager scopeManager,
+                                          BaggageRestrictionManager baggageRestrictionManager,
+                                          boolean expandExceptionLogs,
+                                          JaegerObjectFactory objectFactory) {
+        return new CustomTracer(
+            serviceName,
+            reporter,
+            sampler,
+            registry,
+            clock,
+            metrics,
+            tags,
+            zipkinSharedRpcSpan,
+            scopeManager,
+            baggageRestrictionManager,
+            expandExceptionLogs,
+            objectFactory);
+      }
+    }
+
+    public class CustomSpanBuilder extends JaegerTracer.SpanBuilder {
+      protected CustomSpanBuilder(String operationName) {
+        super(operationName);
+      }
+
+      @Override
+      public CustomSpan start() {
+        return (CustomSpan) super.start();
+      }
+    }
+
+    public CustomTracer(
+        String serviceName,
+        Reporter reporter,
+        Sampler sampler,
+        PropagationRegistry registry,
+        Clock clock,
+        Metrics metrics,
+        Map<String, Object> tags,
+        boolean zipkinSharedRpcSpan,
+        ScopeManager scopeManager,
+        BaggageRestrictionManager baggageRestrictionManager,
+        boolean expandExceptionLogs,
+        JaegerObjectFactory objectFactory) {
+      super(
+          serviceName,
+          reporter,
+          sampler,
+          registry,
+          clock,
+          metrics,
+          tags,
+          zipkinSharedRpcSpan,
+          scopeManager,
+          baggageRestrictionManager,
+          expandExceptionLogs,
+          objectFactory);
+    }
+  }
+
+  private static class CustomSpan extends JaegerSpan {
+    public CustomSpan(
+        JaegerTracer tracer,
+        String operationName,
+        JaegerSpanContext context,
+        long startTimeMicroseconds,
+        long startTimeNanoTicks,
+        boolean computeDurationViaNanoTicks,
+        Map<String, Object> tags,
+        List<Reference> references) {
+      super(
+          tracer,
+          operationName,
+          context,
+          startTimeMicroseconds,
+          startTimeNanoTicks,
+          computeDurationViaNanoTicks,
+          tags,
+          references);
+    }
+
+    @Override
+    public CustomSpanContext context() {
+      return (CustomSpanContext) super.context();
+    }
+  }
+
+  private static class CustomSpanContext extends JaegerSpanContext {
+    public CustomSpanContext(
+        long traceId,
+        long spanId,
+        long parentId,
+        byte flags,
+        Map<String, String> baggage,
+        String debugId,
+        JaegerObjectFactory objectFactory) {
+      super(traceId, spanId, parentId, flags, baggage, debugId, objectFactory);
+    }
+
+    @Override
+    public CustomSpanContext withFlags(byte flags) {
```

Comment:
why do you need to override these?

Reply:
Sadly the covariant return type adds a lot of boilerplate where I have to cast things.

Reply:
For a practical example, here is what happens if I don't override them:

```
CustomSpan customSpan = anotherCustomSpan.withFlags(flags);  // Error: cannot assign JaegerSpan to CustomSpan.
```

It will probably break user code without overriding.

Reply:
Are you worried that someone would write
```
com.uber.jaeger.SpanContext _ = spanCtx.withFlags()
```
?  I highly doubt it, but ok, we can leave these as a form of example.

Reply:
For `SpanContext`, I found 19 cases in internal Uber code. For `Span` I found 281.

Reply:
_referring_ to the class, yes, but I doubt they use these methods. But ok if you want to keep them.

Reply:
I'd just be skeptical about adding them to the actual subclass (not this test)